File: javascript/206.反转链表.js
/*
 * @lc app=leetcode.cn id=206 lang=javascript
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
// 增加一个哨兵节点，进行反转
var reverseList = function(head) {
    if (!head || !head.next) {
        return head;
    };
    // 哨兵节点
    var sentryNode = new ListNode(0);
    sentryNode.next = head;
    var preNode = head; // 开始反转的前置节点
    var currentNode; // 需要前置的节点
    while(preNode.next) {
        currentNode = preNode.next;
        preNode.next = currentNode.next;
        currentNode.next = sentryNode.next;
        sentryNode.next = currentNode;
    };
    return sentryNode.next;
};
// @lc code=end

File: c/206.反转链表.c
/*
 * @lc app=leetcode.cn id=206 lang=c
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* reverseList(struct ListNode* head){
    if (head == NULL || head->next == NULL) {
        return head;
    };
    // 哨兵节点
    struct ListNode sentryNode;
    sentryNode.val = 0;
    sentryNode.next = head;
    struct ListNode* preNode = head; // 开始反转的前置节点
    struct ListNode* currentNode = head->next; // 需要前置的节点
    while(preNode->next != NULL) {
        currentNode = preNode->next;
        preNode->next = currentNode->next;
        currentNode->next = sentryNode.next;
        sentryNode.next = currentNode;
    };
    return sentryNode.next;
}


// @lc code=end

